compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -522,9 +522,13 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
     for ((level, selfSym) <- sortedProxies) {
       lazy val rhsClsSym = selfSym.info.widenDealias.classSymbol
       val rhs =
-        if (lastSelf.exists)
-          ref(lastSelf).outerSelect(lastLevel - level, selfSym.info)
-        else if (rhsClsSym.is(Module) && rhsClsSym.isStatic)
+        if lastSelf.exists then
+          selfSym.info match
+            case info: TermRef if info.isStable =>
+              ref(info)
+            case _ =>
+              ref(lastSelf).outerSelect(lastLevel - level, selfSym.info)
+        else if rhsClsSym.is(Module) && rhsClsSym.isStatic then
```

Comment:
Since we are at the code, I've a question regarding the code just above L511-L517:

```Scala
    val sortedProxies = thisProxy.toList.map {
      case (cls, proxy) =>
        // The class that the this-proxy `selfSym` represents
        def classOf(selfSym: Symbol) = selfSym.info.classSymbol
        // The total nesting depth of the class represented by `selfSym`.
        def outerLevel(selfSym: Symbol): Int = classOf(selfSym).ownersIterator.length
        (outerLevel(cls), proxy.symbol)
    }.sortBy(-_._1)
```

It seems to me that the call `classOf(selfSym)` is useless, as `selfSym` is already a class symbol.

Reply:
The following line (L531) for static objects look dubious, as it is not effective if `lastSelf.exists`:

``` Scala
        else if rhsClsSym.is(Module) && rhsClsSym.isStatic then
```

I expect it to be before the branch  `lastSelf.exists`, thus always generate optimized code.
Currently, it's not the case, as can be seen from the example below:

```Scala
object Wrapper:
  type E <: Throwable
  class C:
    object syntax:
        extension [A <: Matchable](a: E | A)
          transparent inline def foreach(f: A => Any): Unit =
            a match
              case e: E =>
                Wrapper.this.n + m
              case a: A => f(a)
        val m: Int = 4

  def _catch[A](a: => A): E | A =
    try a
    catch case e: E => e

  val n: Int = 3

object throwables extends Wrapper.C
import throwables.syntax.*

@main def Test(): Unit =
  for x <- Wrapper._catch(1/1) do println(x)
```

The proxy generated for `Wrapper` is the following:

```Scala
        val Wrapper$_this: (Wrapper : Wrapper.type) = syntax$_this.2_<outer>
```

And the example crashes the compiler with the current PR.

Reply:
Thanks for these insights!